### PR TITLE
first commit to fish

### PR DESCRIPTION
**Pull request by YugeTen**
1. Added "Fish", as seen in 2021 paper [Gradient matching for domain generalization](https://arxiv.org/pdf/2104.09937.pdf). Included methods:
    - Fish: a first-order algorithm that encourages the learning of invariant features by maximising inter-domain gradient inner product;
    - FishSmooth: a more general version of Fish that offers direct control over the scaling of the gradient inner product term vs. ERM term by specifying `scaling_fish` in hparams.
2. Minor bug fix for class balanced sampling in `domainbed/lib/fast_data_loader.py`

### CODE REVIEW

File: domainbed/algorithms.py [hunk 1]
```diff
@@ -4,15 +4,19 @@
...
+from collections import defaultdict
 
 from domainbed import networks
-from domainbed.lib.misc import random_pairs_of_minibatches
+from domainbed.lib.misc import random_pairs_of_minibatches, ParamDict
 
 ALGORITHMS = [
     'ERM',
+    'Fish',
+    'FishSmooth',
```
Comment (lopezpaz) [inline domainbed/algorithms.py:19]: Hi Yuge! Thank you for your PR. We are currently offering only one version per algorithm. Would you mind removing one fish? This is due to computational budget and to give everyone the same change. You can encapsulate the behaviour of both models using an extra hyper-parameter, if you like.

File: domainbed/algorithms.py [hunk 2]
```diff
@@ -98,6 +102,160 @@ def predict(self, x):
         return self.network(x)
 
 
+class Fish(Algorithm):
+    """
+    Empirical Risk Minimization (ERM)
```
Comment (lopezpaz) [inline domainbed/algorithms.py:107]: Modify docstring with correct name and original reference.

File: domainbed/algorithms.py [hunk 3]
```diff
@@ -98,6 +102,160 @@ def predict(self, x):
...
+        )
+        self.network.reset_weights(meta_weights)
+
+        return {'loss': loss.item()}
+
+    def predict(self, x):
+        return self.network(x)
+
+
+class FishSmooth(Algorithm):
```
Comment (lopezpaz) [inline domainbed/algorithms.py:169]: Remove?

File: domainbed/hparams_registry.py [hunk 4]
```diff
@@ -47,6 +47,12 @@ def _hparam(name, default_val, random_val_fn):
         _hparam('mlp_depth', 3, lambda r: int(r.choice([3, 4, 5])))
         _hparam('mlp_dropout', 0., lambda r: r.choice([0., 0.1, 0.5])) 
 
+    elif algorithm == 'Fish':
+        _hparam('meta_lr', 0.5, lambda r:r.choice([0.05, 0.1, 0.5]))
+
+    elif algorithm == 'FishSmooth':
```
Comment (lopezpaz) [inline domainbed/hparams_registry.py:53]: Remove?

File: domainbed/lib/fast_data_loader.py [hunk 5]
```diff
@@ -16,7 +16,7 @@ class InfiniteDataLoader:
     def __init__(self, dataset, weights, batch_size, num_workers):
         super().__init__()
 
-        if weights:
+        if weights != None:
```
Comment (lopezpaz) [inline domainbed/lib/fast_data_loader.py:19]: replace by `if weights is not None:`

File: domainbed/networks.py [hunk 6]
```diff
@@ -210,3 +211,24 @@ def Classifier(in_features, out_features, is_nonlinear=False):
             torch.nn.Linear(in_features // 4, out_features))
     else:
         return torch.nn.Linear(in_features, out_features)
+
+
+class Whole(nn.Module):
```
Comment (lopezpaz) [inline domainbed/networks.py:216]: Why is this class necessary? Wouldn't a `nn.Sequential` declared inside `Fish` incorporate all these methods?

**Review by lopezpaz — COMMENTED** [review]
First review.

**Comment by lopezpaz** [comment]
Also, please add your algorithm/paper to README.md :)

File: domainbed/networks.py (hunk 6 above)
Comment (YugeTen) [reply domainbed/networks.py:216]: Hi David! Thanks for the quick response. This class is necessary because we need the `reset_weights` method during the training of Fish to apply new weights to the model. It also makes creating a clone of the meta-model without copying the gradient much easier.

If you think this is too clunky I can try to work around it and encapsulate as much into the Fish class as possible.

Comment (lopezpaz) [reply domainbed/networks.py:216]: Sounds good! Could we call it `WholeFish` or something like that? Thanks!

File: domainbed/hparams_registry.py (hunk 4 above)
Comment (lopezpaz) [reply domainbed/hparams_registry.py:53]: Ping to remove this :)

File: domainbed/networks.py (hunk 6 above)
Comment (YugeTen) [reply domainbed/networks.py:216]: Ok! Just to double-check, I can leave the class as it is and rename it as `WholeFish`?

Comment (lopezpaz) [reply domainbed/networks.py:216]: Yes, we can leave it like that. I will look into sub-classing for Fish and ARM in the near future.

**Review by lopezpaz — APPROVED** [review]
Thank you for contributing your great algorithm Fish to DomainBed, Yuge!
